write.go

```diff
@@ -295,19 +279,85 @@ func writeProduceRequestV2(w *bufio.Writer, codec CompressionCodec, correlationI
 	// partition array
 	writeArrayLen(w, 1)
 	writeInt32(w, partition)
+
 	writeInt32(w, size)
+	for _, msg := range msgs {
+		writeMessage(w, msg.Offset, attributes, msg.Time, msg.Key, msg.Value)
+	}
+	return w.Flush()
+}
 
-	if hasHeaders(msgs...) {
-		err = writeRecordBatch(w, codec, correlationID, clientID, topic, partition, timeout, requiredAcks, msgs...)
-	} else {
-		for _, msg := range msgs {
-			writeMessage(w, msg.Offset, attributes, msg.Time, msg.Key, msg.Value)
+func writeProduceRequestV3(w *bufio.Writer, codec CompressionCodec, correlationID int32, clientID, topic string, partition int32, timeout time.Duration, requiredAcks int16, msgs ...Message) (err error) {
+
+	var size int32
+	var compressed []byte
+	var attributes int16
+	if codec != nil {
+		attributes = int16(codec.Code())
+		recordBuf := &bytes.Buffer{}
+		recordBuf.Grow(int(recordBatchSize(msgs...)))
+		compressedWriter := bufio.NewWriter(recordBuf)
+		for i, msg := range msgs {
+			writeRecord(compressedWriter, 0, msgs[0].Time, int64(i), msg)
 		}
+		compressedWriter.Flush()
+
+		compressed, err = codec.Encode(recordBuf.Bytes())
+		if err != nil {
+			return
+		}
+		attributes = int16(codec.Code())
+		size = recordBatchHeaderSize() + int32(len(compressed))
+	} else {
+		size = recordBatchSize(msgs...)
+	}
+
+	h := requestHeader{
+		ApiKey:        int16(produceRequest),
+		ApiVersion:    int16(v3),
+		CorrelationID: correlationID,
+		ClientID:      clientID,
+	}
+	h.Size = (h.size() - 4) +
+		2 + // transactional_id
+		2 + // required acks
+		4 + // timeout
+		4 + // topic array length
+		sizeofString(topic) + // topic
+		4 + // partition array length
+		4 + // partition
+		4 + // message set size
+		size
+
+	h.writeTo(w)
+	writeInt16(w, -1)           // null transactional_id
+	writeInt16(w, requiredAcks) // required acks
+	writeInt32(w, milliseconds(timeout))
+
+	// topic array
+	writeArrayLen(w, 1)
+	writeString(w, topic)
+
+	// partition array
+	writeArrayLen(w, 1)
+	writeInt32(w, partition)
 
+	writeInt32(w, size)
```

Comment:
I'm growing more and more fond of the idea of having a sub-package for all protocol structs so we can try to better control this growing complexity of inlined serializers... work for another day!

Reply:
Totally.  It's only going to get worse now that we have API versions. 😄